codegen/src/main/java/software/amazon/awssdk/codegen/poet/builder/BaseClientBuilderClass.java

```diff
@@ -784,14 +785,32 @@ private MethodSpec internalPluginsMethod() {
         builder.addStatement("$T internalPlugins = new $T<>()", parameterizedTypeName,  ArrayList.class);
 
         for (String internalPlugin : internalPlugins) {
-            ClassName pluginClass = ClassName.bestGuess(internalPlugin);
-            builder.addStatement("internalPlugins.add(new $T())", pluginClass);
+            String arguments = internalPluginNewArguments(internalPlugin);
+            String internalPluginClass = internalPluginClass(internalPlugin);
+            ClassName pluginClass = ClassName.bestGuess(internalPluginClass);
+            builder.addStatement("internalPlugins.add(new $T($L))", pluginClass, arguments);
```

Comment:
What is the purpose of these changes? Can you point me to where it's being used?

Reply:
The purpose of the change is allow internal plugins to take a dependency on `SdkClientConfiguration`. The internal plugins method was changed as

```java
    private List<SdkPlugin> internalPlugins(SdkClientConfiguration config) {
        ⋮
    }
```
If the internal plugin does not care for the config then it should be fine, but if it does (by adding `(config)`) to the customization then it will pass down the config to the constructor. For S3 this is currently codegen as:

```java
    private List<SdkPlugin> internalPlugins(SdkClientConfiguration config) {
        List<SdkPlugin> internalPlugins = new ArrayList<>();
        internalPlugins.add(new S3DisableChunkEncodingIfConfiguredPlugin((config)));
        internalPlugins.add(new S3ExpressPlugin());
        return internalPlugins;
    }
```

The idea behind this is that we want to avoid the cost of wrapping the auth scheme provider (done by  `S3DisableChunkEncodingIfConfiguredPlugin`) for those customers that don't care about it (i.e., they didn't change `chunkedEncodingEnabled` in the `S3Configuration` instance). For that case the plugin should do nothing (i.e., I won't configure the auth scheme provider), but if the setting is set and is not its default value then we will install the wrapping auth-scheme provider. The plugins themselves don't have access to `SdkClientConfiguration` so in order to get access to it the configuration object we need and interceptor + a plugin. 

The constructor of the plugin does this:

```java
    public S3DisableChunkEncodingIfConfiguredPlugin(SdkClientConfiguration config) {
        S3Configuration serviceConfiguration =
            (S3Configuration) config.option(SdkClientOption.SERVICE_CONFIGURATION);

        boolean isServiceConfigurationPresent = serviceConfiguration != null;
        boolean shouldAddDisableChunkEncoding = false;
        boolean isChunkedEncodingEnabledConfigured = false;
        boolean isChunkedEncodingEnabledDisabled = false;
        boolean configuresDisableChunkEncoding = false;
        if (isServiceConfigurationPresent) {
            isChunkedEncodingEnabledConfigured = serviceConfiguration.toBuilder().chunkedEncodingEnabled() != null;
            isChunkedEncodingEnabledDisabled = !serviceConfiguration.chunkedEncodingEnabled();
            configuresDisableChunkEncoding = isChunkedEncodingEnabledConfigured && isChunkedEncodingEnabledDisabled;
            if (configuresDisableChunkEncoding) {
                shouldAddDisableChunkEncoding = true;
            }
        }
        this.configuresDisableChunkEncoding = shouldAddDisableChunkEncoding;
        this.isChunkedEncodingEnabledConfigured = isChunkedEncodingEnabledConfigured;
        this.isChunkedEncodingEnabledDisabled = isChunkedEncodingEnabledDisabled;
        this.isServiceConfigurationPresent = isServiceConfigurationPresent;
    }
```

And then the `configureClient` checks for the `configuresDisableChunkEncoding` flags and sets the wrapper accordingly, e.g.,

```java
    @Override
    public void configureClient(SdkServiceClientConfiguration.Builder config) {
        if (configuresDisableChunkEncoding) {
            LOG.debug(() -> String.format("chunkedEncodingEnabled was explicitly disabled in the configuration, adding "
                                          + "`S3DisableChunkEncodingAuthSchemeProvider` auth provider wrapper."));
            S3ServiceClientConfiguration.Builder s3Config = (S3ServiceClientConfiguration.Builder) config;

            // We wrap the default S3AuthSchemeProvider with a S3DisableChunkEncodingAuthSchemeProvider
            // instance that sets the AwsV4FamilyHttpSigner.CHUNK_ENCODING_ENABLED signer
            // property to false but only if the operations is `"PutObject" or "UploadPart"`
            // This legacy logic was implemented before using an interceptor but now requires
            // wrapping the S3AuthSchemeProvider for it to work.
            S3AuthSchemeProvider disablingAuthSchemeProvider =
                S3DisableChunkEncodingAuthSchemeProvider.create(s3Config.authSchemeProvider());
            s3Config.authSchemeProvider(disablingAuthSchemeProvider);
        } else {
            LOG.debug(() -> String.format("chunkedEncodingEnabled was not explicitly disabled in the configuration, not adding "
                                          + "the `S3DisableChunkEncodingAuthSchemeProvider` auth provider wrapper."
                                          + "[isServiceConfigurationPresent: %s, isChunkedEncodingEnabledConfigured: %s,  "
                                          + "isChunkedEncodingEnabledDisabled: %s]",
                                          isServiceConfigurationPresent,
                                          isChunkedEncodingEnabledConfigured,
                                          isChunkedEncodingEnabledDisabled));
        }
    }
```

Let me know if you know of a better way to accomplish this task.

Reply:
seems fine for now